docs/file-formats/colors.md

```diff
@@ -2,26 +2,123 @@
+| `id`       | `string`  | Yes                                | The unique `id` of the color, which will be used internally by Lona Studio and in the generated code. This should be _code-friendly_: it must not contain spaces or unusual characters, since it will be used as a variable name. |
+| `name`     | `string`  | Yes                                | The human-readable name of the color. This will be displayed in the Lona Studio UI. This will not appear in the generated code.                                                                                                   |
+| `value`    | `object`  | Required if there is no `variants` | An object representing the color. See [Color Value](#color-value) for more information about the properties of the object.                                                                                                        |
+| `variants` | `Color[]` | Required if there is no `value`    | An array representing the variants of the color.                                                                                                                                                                                  |
+| `comment`  | `string`  | No                                 | An optional description of the color, explaining contextual information, such as how it should be used. This may be displayed in the Lona Studio UI. This will not appear in the generated code.                                  |
+
+> The reason why it's an array instead of an object with the id as keys is that we want to order colors, and it's really not recommended to rely on the order of the keys of a JSON object.
+
+### Color Value
+
+#### Value
+
+The value of a color is represented by a JSON object. There are multiple ways to represent it, hence there are multiple set of properties available to represent the color.
+
+- `hex`: `{ "hex": "#00FF00" }`
+- `rgb`: `{ "r": 0, "g": 1, "b": 0 }`
+- `hsl`: `{ "h": 120, "s": 1, "l": 0.5 }`
+- `hsv`: `{ "h": 120, "s": 1, "v": 1 }`
+
+Only one set of properties must be specified for a single value.
+
+#### Aliasing or extending a color
+
+Additionally, it is possible for a value to reference another color's value by their `id` (a `string`) and extend it. (Variants can be referenced by their "id path" (eg. `"colorId.variantId"`))
+
+```json
+{
+  "extends": "anotherColorId"
+}
+```
+
+Once a value extends another one, it can specified any subset of one of the set of properties used to represent a color.
+
+For Example:
+
+```json
+[
+  {
+    "id": "baseColor",
+    "name": "Base Color",
+    "value": { "hex": "#00FF00" }
+  }
+  {
+    "id": "desaturatedBaseColor",
+    "name": "Desaturated Base Color",
+    "value": {
+      "extends": "baseColor",
+      "s": 0.5
```

Comment:
Ooh 👍 for color adjustments

I might even say, let's take it a step further and allow any Lona expression (even if we only implement the very basics to begin with). That'd make it pretty easy to generalize to aliases, color adjustments, or more advanced things potentially.

```swift
struct Color {
  indirect enum Value {
    case color(String)
    case group([Color])
    case ref(LonaExpression)
  }
}
```

Where `LonaExpression` can be things like:

```js
// An identifier, for aliasing
"case": "ref",
"value": "baseColor"

// A member expression, for aliasing (could be dot-separated instead, if every member is an identifier)
"case": "ref",
"value": ["blues", "100"]

// A function call expression (idk exactly the format, but something like)
"case": "ref",
"value": {
  "type": "call",
  "id": "saturate",
  "arguments": ["baseColor", "20"]
}
```

We could optionally have a more concise format for simple/common adjustments. 

```json
"case": "saturate",
"value": {
  "base": "baseColor",
  "s": "20"
}
```

For complicated expressions, e.g. nested adjustment functions, people would definitely want to make them in LS, so I think it's OK if the JSON gets complicated too

In the future, it'd be cool if we could even generate a whole palette or group of colors dynamically. Like, palette generations plugins

Reply:
I like that! Where would the functions be declared? Are they pre-defined?

Reply:
I think pre-defined to start, and then maybe at some point we can make them extendable